envision/server.py

```diff
@@ -54,30 +64,201 @@ def options(self):
         self.finish()
 
 
+class Frame:
+    def __init__(self, data: str, timestamp: int, next_=None):
+        # Time since epoch in seconds
+        self._timestamp = timestamp
+        self._data = data
+        self._size = sys.getsizeof(data)
+        self.next_ = next_
+
+    @property
+    def timestamp(self):
+        return self._timestamp
+
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, data: str):
+        self._data = data
+        self._size = sys.getsizeof(data)
+
+    @property
+    def size(self):
+        return self._size
+
+
+class Frames:
+    def __init__(self, max_capacity_mb=500):
+        self._max_capacity = max_capacity_mb
+
+        # XXX: Index of timestamp in `self._timestamps` matches the index of the
+        # frame with the same timestamp in `self._frames`.
+        self._timestamps = []
+        self._frames = []
+
+    @property
+    def start_frame(self):
+        if len(self._frames) == 0:
+            return None
+
+        return self._frames[0]
+
+    @property
+    def start_time(self):
+        return self._frames[0].timestamp if self._frames else None
+
+    @property
+    def elapsed_time(self):
+        if len(self._frames) == 0:
+            return 0
+
+        return self._frames[-1].timestamp - self._frames[0].timestamp
+
+    def append(self, frame: Frame):
+        self._enforce_max_capacity()
+
+        if len(self._frames) >= 1:
+            self._frames[-1].next_ = frame
+
+        self._frames.append(frame)
+        self._timestamps.append(frame.timestamp)
+
+    def __call__(self, timestamp):
+        """Finds the nearest frame according to the given timestamp."""
+        frame_idx = bisect.bisect_left(self._timestamps, timestamp)
+        if frame_idx >= len(self._frames):
+            frame_idx = -1
+
+        return self._frames[frame_idx]
+
+    def _enforce_max_capacity(self):
+        """Sample random frames and clear their data to ensure we're under the max
+        capacity size.
+        """
+        bytes_to_mb = 1e-6
+        sizes = [frame.size for frame in self._frames]
+        while len(self._frames) >= 2 and sum(sizes) * bytes_to_mb > self._max_capacity:
+            # XXX: randint(1, ...), we skip the start frame because that is a "safe"
+            #      frame we can always rely on being available.
+            idx_to_delete = random.randint(1, len(self._frames) - 1)
```

Comment:
I still have some concern that it might be best to exclude a number of the newest frames so that the most recent frames will be guaranteed to be viewable.

Reply:
Sure, doesn't hurt. I'll do that now.